akka-typed/src/test/scala/docs/akka/typed/IntroSpec2.scala

```diff
@@ -0,0 +1,144 @@
+    final case class Greeted(whom: String)
+
+    // TODO Is the small difference between Stateless and Stateful worth separate classes?
+    //      Stateful returning Same is the same as Stateless
+    val greeter: Behavior[Greet] = Stateless { (_, msg) ⇒
+      println(s"Hello ${msg.whom}!")
+      msg.replyTo ! Greeted(msg.whom)
+    }
+  }
+  //#hello-world-actor
+
+  //#chatroom-actor
+  object ChatRoom {
+    //#chatroom-protocol
+    sealed trait Command
+    final case class GetSession(screenName: String, replyTo: ActorRef[SessionEvent])
+      extends Command
+    //#chatroom-protocol
+    //#chatroom-behavior
+    private final case class PostSessionMessage(screenName: String, message: String)
+      extends Command
+    //#chatroom-behavior
+    //#chatroom-protocol
+
+    sealed trait SessionEvent
+    final case class SessionGranted(handle: ActorRef[PostMessage]) extends SessionEvent
+    final case class SessionDenied(reason: String) extends SessionEvent
+    final case class MessagePosted(screenName: String, message: String) extends SessionEvent
+
+    final case class PostMessage(message: String)
+    //#chatroom-protocol
+    //#chatroom-behavior
+
+    val behavior: Behavior[GetSession] = {
+      var sessions = List.empty[ActorRef[SessionEvent]]
+
+      // TODO it's not obvious that this type annotation is needed, is there a better way?
+      val b: Behavior[Command] =
+        Stateful {
+          case (ctx, GetSession(screenName, client)) ⇒
+            sessions ::= client
+            val wrapper = ctx.spawnAdapter {
+              p: PostMessage ⇒ PostSessionMessage(screenName, p.message)
+            }
+            client ! SessionGranted(wrapper)
+            Same
+          case (ctx, PostSessionMessage(screenName, message)) ⇒
+            val mp = MessagePosted(screenName, message)
+            sessions foreach (_ ! mp)
+            Same
+        }
+      b.narrow // only expose GetSession to the outside
+      //#chatroom-behavior
+    }
+  }
+  //#chatroom-actor
+
+}
+
+class IntroSpec2 extends TypedSpec {
+  import IntroSpec2._
+
+  def `must say hello`(): Unit = {
+    //#hello-world
+    import HelloWorld._
+    // TODO this is not something we would like to encourage and have in docs
+    // using global pool since we want to run tasks after system.terminate
+    import scala.concurrent.ExecutionContext.Implicits.global
+
+    val system: ActorSystem[Greet] = ActorSystem("hello", greeter)
+
+    val future: Future[Greeted] = system ? (Greet("world", _))
+
+    for {
+      greeting ← future.recover { case ex ⇒ ex.getMessage }
+      done ← { println(s"result: $greeting"); system.terminate() }
+    } println("system terminated")
+    //#hello-world
+  }
+
+  def `must chat`(): Unit = {
+    //#chatroom-gabbler
+    import ChatRoom._
+
+    val gabbler: Behavior[SessionEvent] =
+      Stateful { (_, msg) ⇒
+        msg match { // TODO what happens with unhandled messages when written like this?
```

Comment:
MatchError, followed by actor termination. Which is why Scala will rightfully warn—if the message type is sealed.